tests: Add added base RPMI tests

slingappa opened the pull request:
Added Base service group tests.
Refer to test/README.md for mode details

avpatel commented on Makefile:
```diff
@@ -102,6 +102,8 @@ ARFLAGS		=	rcs
 ELFFLAGS	+=	-static
 ELFFLAGS	+=	$(EXTRA_ELFFLAGS)
 
+LDFLAGS	+= -L$(build_dir) -lrpmi
```
Add these to `ELFFLAGS` just after `-static`

avpatel replied:
Also, the top-level Makefile change should be a separate patch.

slingappa replied:
Done

avpatel commented on include/librpmi_env.h:
```diff
@@ -59,6 +59,32 @@ typedef long			rpmi_ssize_t;
+ * @param[in] n		number of bytes to compare
+ * @return void *	Pointer to destination
+ */
+static inline int rpmi_env_memcmp(void *s1, void *s2, rpmi_size_t n)
```
The change in librpmi_env.h should be a separate patch.

avpatel commented on test/test_shmem.c:
```diff
@@ -0,0 +1,155 @@
+static enum rpmi_error shmem_env_memcpy_read(void *priv, rpmi_uint64_t addr,
+					     void *in, rpmi_uint32_t len)
+{
+	rpmi_env_memcpy(in, (void *)addr, len);
```
Never directly case integer to pointer. Instead first cast it to `unsigned long` and then cast it to pointer.

For example: `rpmi_env_memcpy(in, (void *)(unsigned long)addr, len);`

avpatel replied:
This is to avoid compilation errors for 32bit systems.

slingappa replied:
This file is gone now.

slingappa commented:
@avpatel took care of the comments, please review

avpatel commented on test/README.md:
```diff
@@ -0,0 +1,74 @@
+
+## 2.1 Running qemu
+
+Execute qemu instance with shared memory backend for RAM:
```
These tests are mock test which we can directly run as application on x86 host so no need to launch QEMU.

slingappa replied:
can be yes, but some has to write the server if they're not using the qemu right ?

avpatel replied:
better to have both client and server in the same test

avpatel replied:
Basically each test should define the messages exchanged client and server. The test fails if expected response is not received a particular message.

avpatel replied:
One test can be something like this:

1. Enqueue a request (message)
2. Process the context to get response
3. Dequeue the response (message)
4. Inspect the response to declare pass/fail

pathakraul replied:
If possible, a summary of all tests passed/failed too

slingappa replied:
These are already done:
- Basically each test should define the messages exchanged client and server. The test fails if expected response is not received a particular message.

- a summary of all tests passed/failed too, in the end , please refer to README.md inside test folder for sample output

slingappa replied:
> better to have both client and server in the same test

Server code inside librpmi would bloat it, as we'll have to paste all of the qemu RPMI implentation in the server code. better to point them to upstream qemu branch which already has all of it and provide samepl qemu command , like we have in the README file no ?

slingappa replied:
> One test can be something like this:
> 
> 1. Enqueue a request (message)
> 2. Process the context to get response
> 3. Dequeue the response (message)
> 4. Inspect the response to declare pass/fail

This is done in _base* file, please check

avpatel replied:
Okay, sounds good.

But, we still don't need to launch QEMU for running librpmi test applications.

slingappa replied:
We do, for now base group might work with small snippet to initialize that, but for other groups like HSM, CPPC and all still needs to form the response another server app still needs to emulate, which essentially is being done in qemu.

If we dont want to do that, then, we'll have to do one of these:
1. Use the running qemu instance
2. write a test server app inside qemu code which links all the rpmi code
3. write a test server app out side qemu copy all the rpmi code.
So I still think option#1 above seems simplest one from users point of view.

Let me know if you guys would like to implement any one of teh above still or any other server emulation, will do it.

pathakraul replied:
I thought that our current shmem(transport) can be entirely mocked on the x86 host? 
Both sides of transport (server and client) can be invoked in sequential manner to make one complete message request response transaction. Bigger test functions can be written which will comprise multiple such function calls in sequence of both sides(server and client)

I am just thinking from a user perspective where i dont want to have qemu launch which will require me to first build it and also find the right repo and things which go with it.

Thoughts?

slingappa replied:
> I thought that our current shmem(transport) can be entirely mocked on the x86 host?
> 
> Both sides of transport (server and client) can be invoked in sequential manner to make one complete message request response transaction. Bigger test functions can be written which will comprise multiple such function calls in sequence of both sides(server and client)
> 
> I am just thinking from a user perspective where i dont want to have qemu launch which will require me to first build it and also find the right repo and things which go with it.
> 
> Thoughts?

ofcourse, it can be mocked, as mentioned in options 2 or 3 [here](https://github.com/riscv-software-src/librpmi/pull/19#discussion_r1690923413). 

But.. see my concerns mentioned in the same comment. if we really want to we can copy qemu implementation for current and future service groups. I personally dont think its a good idea to push an test implementation which bloats up librpmi.

avpatel commented:
Change the patch subject of last commit from `tests: Add added base RPMI tests` to `test: Add base RPMI tests`

slingappa commented:
> test: Add base RPMI tests

done

slingappa commented:
@avpatel @pathakraul ,
Added two more commits, which implements a test server to test Base group.

Now, no need to qemu, librpmi can be (mock)tested on any linux machine.

Again I'll let you guys to decide, 
- if and when we do decide this is what we need we can pick all these 5 commits,
- if we want to roll with qemu, then omit last 2 commits.

avpatel commented on test/objects.mk:
```diff
@@ -3,13 +3,22 @@
 # Copyright (c) 2024 Ventana Micro Systems Inc.
 #
+#
+test-elfs-y += test_srv
```
I still don't understand. Why separate server application ?

With this approach, each test scenario will require a separate test application.

slingappa replied:
No, it doesnt, 

` process_ap_reqs()` inside `test_srv.c ` is common for all groups,
but server implementation definitely needs to be added just like we did in qemu for each service groups seperately as depicted below in  `add_*_group()` below:

```
 int main(int argc, char *argv[])
 {
         printf("Test server thread...\n");

         /*
          * along with contexts only BASE group is created, rest needs to be
          * added here later, after the context creation
          */
         create_contexts();
         
         add_hsm_group(&hsm);
         add_cppc_group(&cppc)
         
         process_ap_reqs();

         return 0;
 }

```

avpatel replied:
This approach of adding a separate server for various scenarios is not going to scale. We have to re-design the common test infrastructure.

avpatel commented on test/test_common.c:
```diff
@@ -5,3 +5,148 @@
+	printf("%-50s \t : %s!\n", rpmi_test->name, failed?"Failed":"Succeeded");
+}
+
+void execute_rpmi_test_group(struct rpmi_transport *rpmi_xport, struct rpmi_test_group *group)
```
We can improve this common test infrastructure such that each test has both client and server part of the test-case. This will help us avoid creating a separate server application for each test scenario.

avpatel commented on test/test_common.h:
```diff
@@ -7,4 +7,90 @@
+
+struct rpmi_test_group {
+	char name[64];
+	int num_tests;
```
additonal members

rpmi_shmem *shmem;
rpmi_transport *trans;
rpmi_context *context;

avpatel commented on test/test_common.h:
```diff
@@ -7,4 +7,90 @@
+	char name[64];
+	int num_tests;
+	struct rpmi_test tests[];
+};
```
setup()
cleanup()

slingappa commented:
@avpatel @pathakraul ,

please check the changes now modified as per our discussion.

users are expected to modify `struct rpmi_test_group` members to create different test scenarios. and run `execute_test_group` repeatedly.

avpatel commented on test/test_common.h:
```diff
@@ -7,4 +7,126 @@
+ * its better to decode these data from DTS file, hardcoding for now.
+ */
+
+#define RPMI_SHM_XPORTS       (2)
```
Drop this.

slingappa replied:
agreed, dropping.

avpatel commented on test/test_common.h:
```diff
@@ -7,4 +7,126 @@
+ * test assums qemu RPMI shared memory address is mapped at the following
+ * offset in gues RAM
+ */
+#define RPMI_SHM_OFSET       (0x0000000010001000)
```
We don't need lot of these defines.

slingappa replied:
agreed, was needed for qemu, not anymore dropping.

avpatel commented on test/test_common.h:
```diff
@@ -7,4 +7,126 @@
+
+struct rpmi_test {
+	char name[64];
+	void *rpmi_xport;
```
No need for this pointer. See below.

avpatel commented on test/test_common.h:
```diff
@@ -7,4 +7,126 @@
+	unsigned char resp_data[RPMI_MSG_MAX_DATA_SIZE];
+
+	/* test specific function handlers*/
+	int (*init)(struct rpmi_test *test);
```
For all the test callbacks, just pass rpmi_test_group pointer which already has rpmi_context and rpmi_transport pointers.

slingappa replied:
No, rpmi_test still needs as we still need to track, request/response data per service group tests.

avpatel commented on test/test_transport.c:
```diff
@@ -0,0 +1,155 @@
+		printf("data[%2d]: %08x \n", i, data[i]);
+}
+
+void *get_shm_addr(unsigned int xport_id, unsigned int *shm_sz)
```
We don't need this anymore.

avpatel commented on test/test_transport.c:
```diff
@@ -0,0 +1,155 @@
+		free(msg);
+}
+
+struct rpmi_transport *init_test_rpmi_xport(int xport_id, int do_clear)
```
This function is not need. Most of this things can be done in init_group().

In fact, you can move important stuff from test_transport.c and remove test_transport.c

avpatel commented on test/test_common.c:
```diff
@@ -5,3 +5,237 @@
+void cleanup_group(struct rpmi_test_group *group)
+{
+	free(group->xport->priv);
+	free(group->rctx);
```
Use librpmi APIs.

slingappa replied:
Done

slingappa commented:
@avpatel 
- `test_common.c` allocates/frees for every test memory instead of opening `/dev/shm/` file
- removed `*transport.c` file and brought in required changes inside `common.c` file
- removed multi transport related changes, so `xport_id` gone from everywhere.

avpatel commented on include/librpmi_env.h:
```diff
@@ -368,7 +394,10 @@ static inline rpmi_uint32_t rpmi_to_xe32(rpmi_bool_t is_be, rpmi_uint32_t val)
  * @param[in] val	4-byte data to write
  */
-void rpmi_env_writel(rpmi_uint64_t addr, rpmi_uint32_t val);
+static inline void rpmi_env_writel(rpmi_uint64_t addr, rpmi_uint32_t val)
```
This should be implemented in test_common.c

avpatel commented:
@slingappa Much better now. Few comments remaining.
